(PUP-11405) Replace deprecated/changed Psych YAML methods, pin rdoc

tambry opened the pull request:
The non-keyword argument variant was deprecated in Psych 3.1.0 and has been
removed in 4.0.0 (https://github.com/ruby/psych/commit/0767227051dbddf1f949eef512c174deabf22891).

Encountered this in a downstream custom Debian packaging of Puppet server with Psych 4.0.3 and manifested as an error regarding too many arguments (1 vs 5) during global `lookup_options` lookup resulting in catalog compilation failure.

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppet?pullRequest=8849) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppet?pullRequest=8849) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppet?pullRequest=8849) it.</sub>

ekohl reviewed (commented):
I'm also running into this and is blocking our CI pipelines. However, in Ruby 2.5.9 with Psych 3.0.2 this fails:
```console
$ irb
irb(main):001:0> require 'yaml'
=> true
irb(main):002:0> YAML.safe_load('[]', permitted_classes: [], aliases: true, filename: nil)
Traceback (most recent call last):
        3: from /usr/bin/irb:11:in `<main>'
        2: from (irb):2
        1: from /usr/share/ruby/psych.rb:313:in `safe_load'
ArgumentError (unknown keywords: permitted_classes, aliases, filename)
```

tambry commented:
> I'm also running into this and is blocking our CI pipelines. However, in Ruby 2.5.9 with Psych 3.0.2 this fails:

I assumed Puppet didn't support such old versions, but seems Ruby 2.5 is still supported. I've guarded the non-deprecated variant behind a version check.

ekohl commented:
Puppet 6 ships Ruby 2.5 and I doubt it's reasonable to rebase that. The alternative is to add a dependency on Psych < 4 in the gemspec but I think your current solution is cleaner.

ekohl commented:
Looks like the CI pipeline itself also has problems:
```
Run bundle exec rake rubocop
Your RubyGems version (2.7.6.3)) has a bug that prevents `required_ruby_version` from working for Bundler. Any scripts that use `gem install bundler` will break as soon as Bundler drops support for your Ruby version. Please upgrade RubyGems to avoid future breakage and silence this warning by running `gem update --system 3.2.3`
wrong number of arguments (given 5, expected 1)
/opt/hostedtoolcache/Ruby/2.5.9/x64/lib/ruby/gems/2.5.0/gems/psych-4.0.3/lib/psych.rb:323:in `safe_load'
```

tambry commented:
> Looks like the CI pipeline itself also has problems:

Seems to be because the pipeline is now using a newer Psych, but an older Rubocop, which uses the now removed load method, which this PR fixes for Puppet. 😃

ekohl commented:
Yep, psych 4 is breaking things all over the ecosystem. Great start of the new year.

joshcooper commented on lib/puppet/util/yaml.rb:
```diff
@@ -24,7 +24,11 @@ class YamlLoadError < Puppet::Error; end
   # @return The parsed YAML, which can be Hash, Array or scalar types.
   def self.safe_load(yaml, allowed_classes = [], filename = nil)
-    data = YAML.safe_load(yaml, allowed_classes, [], true, filename)
+    if Psych::VERSION >= "3.1.0"
```
Thanks for the contribution @tambry!

Comparing strings can lead to surprises:

```
irb(main):015:0> "3.1" >= "3.1.0"
=> false
```

Might be better off using `to_f`, which we use in other places like `RUBY_VERSION.to_f`:

```ruby
if Psych::VERSION.to_f >= 3.1
```

The psych dependency is getting pulled in due to `rdoc`:

```
    rdoc (6.4.0)
      psych (>= 4.0.0)
```

And apparently `rdoc` doesn't follow SemVer: https://github.com/ruby/rdoc/issues/857#issuecomment-1003787128

Could you update https://github.com/puppetlabs/puppet/blob/6.x/Gemfile#L32 so it excludes 6.0.4, something like:

```
gem 'rdoc', ['~> 6.0', '< 6.0.4'], require: false, platforms: [:ruby]
```

If you'd like attribution in our release notes, please file a PUP ticket and update the commit summary and GH PR title with `(PUP-XXX) Replace  deprecated ...` Thanks!

tambry replied:
Thanks for the review! I've made all the suggested changes.

joshcooper commented:
@tambry Thanks for the update. It looks like this issue affects puppet#6.x too (see https://github.com/puppetlabs/puppet/pull/8850) Could you rebase your feature branch on puppet#6.x, force push and update the base branch for this PR to `6.x`?

Psych also changed the behavior of YAML.load so it's safe by default, which breaks our tests. Those tests will need to be updated to either call the `unsafe_load_file` or pass a list of `permitted_classes`. Probably better to do the latter, so it works on all ruby/psych versions.

Also it looks like GH is using ruby 3.1 which is introducing other issues. I'll take a look at that.

tambry commented:
@joshcooper Done.

> Psych also changed the behavior of YAML.load so it's safe by default, which breaks our tests. Those tests will need to be updated to either call the unsafe_load_file or pass a list of permitted_classes. Probably better to do the latter, so it works on all ruby/psych versions.

I've updated the `load_file()` wrapper to use `unsafe_load_file()` in versions where it's available thus keeping the behaviour. I'm not very experienced with Ruby so I'd prefer to leave cleaning up the tests up to someone else.

luchihoratiu commented:
Trying to reset CI.

luchihoratiu commented on lib/puppet/util/yaml.rb:
```diff
@@ -24,7 +24,11 @@ class YamlLoadError < Puppet::Error; end
   # @return The parsed YAML, which can be Hash, Array or scalar types.
   def self.safe_load(yaml, allowed_classes = [], filename = nil)
-    data = YAML.safe_load(yaml, allowed_classes, [], true, filename)
+    if Psych::VERSION.to_f >= 3.1
```
```suggestion
    if Gem::Version.new(Psych::VERSION) >= Gem::Version.new('3.1.0')
```

luchihoratiu commented on lib/puppet/util/yaml.rb:
```diff
@@ -68,7 +72,11 @@ def self.load_file(filename, default_value = false, strip_classes = false)
       data.to_ruby || default_value
     else
-      yaml = YAML.load_file(filename)
+      if Psych::VERSION.to_f >= 3.3.2
```
```suggestion
    if Gem::Version.new(Psych::VERSION) >= Gem::Version.new('3.3.2')
```

luchihoratiu reviewed (commented):
To get the code working and tests running, we need something like these suggestions.

tambry commented:
> To get the code working and tests running, we need something like these suggestions.

Done.

joshcooper reviewed (approved):
Thanks @tambry for your contribution! This will be released in the next versions in the 6.x and 7.x streams.

Note the `.to_f` trick "works" when comparing just the major and minor versions, for example, we do that in https://github.com/puppetlabs/puppet/blob/3b8b7ad6fcaec50b8b7d4383fac9aea8e75ae560/lib/puppet/util/monkey_patches.rb#L32. But using `Gem::Version` is "more correct", especially for the `3.3.2` case.

ekohl commented:
Thanks everyone! I'd love to get this in a released gem soon to fix our CI pipelines.
